server/hotrod/src/main/scala/org/infinispan/server/hotrod/Decoder2x.scala

```diff
@@ -234,209 +217,185 @@ object Decoder2x extends AbstractVersionedDecoder with ServerConstants with Log
+            } yield {
+               hrCtx.operationDecodeContext = (segments, factory, batchSize, metadata)
+               buffer.markReaderIndex()
+               out.add(hrCtx)
             }
-            val batchSize = readUnsignedInt(buffer)
-            val metadata = Constants.isVersionPost24(h.version) && buffer.readByte() != 0
-            val iterationId = server.iterationManager.start(cache.getName, segments.map(JavaBitSet.valueOf), namedFactory, batchSize, metadata)
-            new IterationStartResponse(h.version, h.messageId, h.cacheName, h.clientIntel, h.topologyId, iterationId)
-         case IterationNextRequest =>
-            val iterationId = readString(buffer)
-            val iterationResult = server.iterationManager.next(cache.getName, iterationId)
-            new IterationNextResponse(h.version, h.messageId, h.cacheName, h.clientIntel, h.topologyId, iterationResult)
-         case IterationEndRequest =>
-            val iterationId = readString(buffer)
-            val removed = server.iterationManager.close(cache.getName, iterationId)
-            new Response(h.version, h.messageId, h.cacheName, h.clientIntel, IterationEndResponse, if (removed) Success else InvalidIteration, h.topologyId)
-         case _ => null
+         case NewHotRodOperation.IterationNextRequest | NewHotRodOperation.IterationEndRequest =>
+            readMaybeString(buffer).foreach(iterationId => {
+               hrCtx.operationDecodeContext = iterationId
+               buffer.markReaderIndex()
+               out.add(hrCtx)
+            })
+         case _ =>
       }
    }
 
-   private def readNamedFactory(buffer: ByteBuf): NamedFactory = {
-      for {
-         factoryName <- readOptionalString(buffer)
-      } yield (factoryName, readOptionalParams(buffer))
+   private def readMaybeNamedFactory(buffer: ByteBuf): Option[NamedFactory] = {
+      readMaybeString(buffer).flatMap(name => {
+         if (!name.isEmpty) {
+            readOptionalParams(buffer).map(param => Some(name, param))
+         } else Some(None)
+      })
    }
 
-   private def readOptionalString(buffer: ByteBuf): Option[String] = {
-      val string = readString(buffer)
-      if (string.isEmpty) None else Some(string)
-   }
-
-   private def readOptionalParams(buffer: ByteBuf): List[Bytes] = {
-      val numParams = buffer.readByte()
-      if (numParams > 0) {
-         var params = ListBuffer[Bytes]()
-         for (i <- 0 until numParams) params += readRangedBytes(buffer)
-         params.toList
-      } else List.empty
+   private def readOptionalParams(buffer: ByteBuf): Option[List[Bytes]] = {
+      val numParams = readMaybeByte(buffer)
+      numParams.map(p => {
+         if (p > 0) {
+            val params = ListBuffer[Bytes]()
+            @tailrec def readParams(buf: ListBuffer[Bytes]): Boolean = {
+               if (readMaybeRangedBytes(buffer).exists(param => {
```

Comment:
Taking in account what I said above WRT sealed traits, I think it'd make this whole section a lot more readable by pattern matching on the 3 possible outcomes.